KAFKA-6895: (WIP) Schema Inferencing for JsonConverter

natengall opened the pull request:
**Work in Progress: Do not merge until KIP and this PR are both approved.**

KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-301%3A+Schema+Inferencing+for+JsonConverter

This introduces a new configuration for the JsonConverter class called "schemas.infer.enable". When "schemas.enable" is false and the "schemas.infer.enable" flag is set to true, the converter will infer the schema from the contents of the JSON record and return that as part of the SchemaAndValue object for Sink Connectors.

Author: Allen Tang <natengall@gmail.com>

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

natengall committed:
Schema Inferencing for JsonConverter

Introduce a new configuration for the JsonConverter class called "schemas.infer.enable". When this flag is set to "false", the existing behavior is exhibited. When it's set to "true", infer the schema from the contents of the JSON record, and return that as part of the SchemaAndValue object for Sink Connectors.

Author: Allen Tang <natengall@gmail.com>

rhauch commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -361,6 +367,22 @@ private SchemaAndValue jsonToConnect(JsonNode jsonValue) {
 
+    private SchemaAndValue schemalessJsonToConnect(JsonNode jsonValue) {
+        if (jsonValue == null)
+            return SchemaAndValue.NULL;
```
Nit: we always use braces around if-blocks.

natengall replied:
(Always)[https://github.com/apache/kafka/blob/trunk/connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java#L354-L355]? ;)

I'll apply that rule for the whole file in a followup commit.

rhauch commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -755,4 +777,40 @@ private static Object convertToConnect(Schema schema, JsonNode jsonValue) {
         Object convert(Schema schema, Object value);
     }
+
+    private void flatten(String key, JsonNode jsonValue, LinkedHashMap<String, Object> map, SchemaBuilder builder) {
```
What happens when you have a simple JSON value like string values (e.g., `"foo"`), `null`, `true`, `false`, numbers (e.g., `1234.5`), or even an array? This method should handle all potential JSON values, not just JSON object values.

To do this correctly, we may need to separate the schema inference from the value construction. And if we introduce a method that infers the `Schema` from the JSON value (and this method could be recursive), then we should also be able to reuse the existing `convertToConnect(Schema schema, JsonNode jsonValue)` method. It's not quite as fast since we're iterating over the value twice, but this is actually how the schema-enabled version of the code works and it would reuse a lot more existing code.

WDYT?

natengall replied:
Thanks for the input, @rhauch!

My focus so far in this initial version of schemaless json support was to enable data ingestion via JDBCSinkConnector without the explicit need for producers to provide schemas. This means only ever resulting in Kafka Connect Structs at the outermost level, which means there's an assumption made that the record is a JSON object. I do agree there is some opportunity to remove code duplication with the existing convertToConnect method. Let me see what I can do to make this more generic for simple JSON values.

rhauch replied:
Thanks for considering my suggestions. Looking forward to the changes!

As for only handling structs, we can't constrain how people use this or what goes into their streams that they consume with a Connect sink using this connector, except to say that this should handle all all possible valid JSON documents, including those comprised of simple JSON values. :-)

rhauch reviewed (commented):
@natengall thanks for the PR. This is a really good first step. I left some comments below to address more JSON value cases and to reuse more of the existing code. We also need test cases that verify the new behavior against a thorough variety of JSON values.

Finally, I know we discussed this, but this is a behavioral change and so we'll need a KIP. Please rename the PR to denote that it is a work in progress, something like:

```
KAFKA-6895: (WIP) Schema Inferencing for JsonConverter
```
and then make sure the description starts with:

```
**Work in Progress: Do not merge until KIP and this PR are both approved.**
```

natengall committed:
Schema Inferencing for JsonConverter
Accept the domain of all JSON documents, including those comprised of simple JSON values

natengall committed:
Schemaless Json Unit Tests
- implemented prepareSchemalessJsonConverter(), a method to prep JsonConverter configs: schemas.enable=false and schemas.infer.enable=true
- implemented unit tests, prefixed by "schemaless", which are modeled after existing tests in JsonConverterTest, but with test cases that do not contain JSON envelopes with schema data: schemalessArrayToConnect, schemalessBooleanToConnect, schemalessDoubleToConnect, schemalessLongToConnect, schemalessStringToConnect, schemalessStructToConnect

rhauch commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverterConfig.java:
```diff
@@ -39,6 +39,11 @@
 
+    public static final String SCHEMAS_INFER_ENABLE_CONFIG = "schemas.infer.enable";
+    public static final boolean SCHEMAS_INFER_ENABLE_DEFAULT = false;
+    private static final String SCHEMAS_INFER_ENABLE_DOC = "Infer the schemas when they are missing within each of the serialized values and keys.";
```
Might be useful to mention the serialized representation is JSON, and to say that this only affects deserialization when `schemas.enable=false`.

rhauch commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -755,4 +770,35 @@ private static Object convertToConnect(Schema schema, JsonNode jsonValue) {
+                    return Schema.FLOAT64_SCHEMA;
+                }
+            case ARRAY:
+                SchemaBuilder arrayBuilder = SchemaBuilder.array(inferSchema(jsonValue.elements().next()));
```
What happens if the array is empty?

rhauch commented on connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java:
```diff
@@ -139,12 +165,25 @@ public void stringToConnect() {
+        prepareSchemalessJsonConverter();
+        byte[] arrayJson = "[1, 2, 3]".getBytes();
+        assertEquals(new SchemaAndValue(SchemaBuilder.array(Schema.INT64_SCHEMA).build(), Arrays.asList(1L, 2L, 3L)), converter.toConnectData(TOPIC, arrayJson));
+    }
```
How about testing for an empty array?

natengall replied:
Good catch! I'm making the inference that empty JSON arrays contain an array of strings. Let me know your thoughts on this.

rhauch commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -755,4 +770,35 @@ private static Object convertToConnect(Schema schema, JsonNode jsonValue) {
+            case STRING:
+                return Schema.STRING_SCHEMA;
+            default:
+                return null;
```
Jackson has several other custom node types: `BINARY`, `MISSING`, and `POJO`. We should cover all possible options with additional case statements, like on [lines 729-732](https://github.com/apache/kafka/blob/1.1.0/connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java#L729-L732).

natengall replied:
Thanks for the suggestion. I extended the behavior from lines 729-732, explicitly listing out the three json node types to return null schemas here.

rhauch reviewed (commented):
@natengall looks great! Just a few pretty minor comments/suggestions.

natengall committed:
Support inferencing of empty json arrays
- Handle empty json array by inferring an array of optional strings
- Explicitly list BINARY, MISSING, and POJO json node types in inferencing switch-case

rhauch commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -785,7 +785,7 @@ private Schema inferSchema(JsonNode jsonValue) {
                 }
             case ARRAY:
-                SchemaBuilder arrayBuilder = SchemaBuilder.array(inferSchema(jsonValue.elements().next()));
+                SchemaBuilder arrayBuilder = SchemaBuilder.array(jsonValue.elements().hasNext() ? inferSchema(jsonValue.elements().next()) : Schema.OPTIONAL_STRING_SCHEMA);
```
It might be better to pull the inference out to different line, since it's possible that the `inferSchema` method returns null, and then handle that null case as `SchemaBuilder.array(...)` does not tolerate a null schema.

Regardless, this is just one of the assumptions that will have to be documented in the KIP. I suspect these assumptions are where most of the contention will be. It's not clear that assuming an empty array is a String array is necessarily the best option.

rhauch commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverterConfig.java:
```diff
@@ -41,7 +41,7 @@
     public static final String SCHEMAS_INFER_ENABLE_CONFIG = "schemas.infer.enable";
     public static final boolean SCHEMAS_INFER_ENABLE_DEFAULT = false;
-    private static final String SCHEMAS_INFER_ENABLE_DOC = "Infer the schemas when they are missing within each of the serialized values and keys.";
+    private static final String SCHEMAS_INFER_ENABLE_DOC = "Infer the schemas when they are missing within each of the serialized values and keys. Only applied when " + SCHEMAS_ENABLE_CONFIG + " is false.";
```
I might suggestion:

    "Infer the schemas when deserializing JSON and when " + SCHEMAS_ENABLE_CONFIG + " is false and the JSON keys/values have no embedded schemas."

rhauch reviewed (commented):
One more pass. We need to start discussing the higher level constraints and assumptions on the KIP's discussion thread.

ottomata commented:
Thought: even better than inferring schema, would be a pluggable way to set the schema, of which inferring the schema would be one.  I'd use something like this to get a real JSONSchema for each record, and then convert the JSONSchema into a Connect Schema.

helpermethod commented:
Any progress on this? This looks like an awesome addition for those using JsonConverter in combination with JDBC Sink Connector.

helpermethod commented:
Hi! How are you planing to handle timestamps? For the algorithm it's probably impossible to decide if an int64 is supposed to be a timestamp or not, or am I wrong?

Whenusing the sink connector to ingest data from another source it's likely that the payload contains timestamp fields. Which would need additional meta information to be recognized as such.

ottomata commented:
FYI: https://github.com/ottomata/kafka-connect-jsonschema is WIP, but set's the schema using a JSONSchema rather than inferring it from JSON data.

helpermethod commented:
@ottomata Thank you, very interesting project! But it doesn't quite answer my question: Will the Schema Inferencing for the JSON Converter be able to infer Timestamps?

ottomata commented:
Oh, sorry, @helpermethod  my comment was meant in reference to my previous one about how it would be nice if this task about about implementing a more generic configuration for getting the schema for a topic, of which JSON inference or my JSONSchema stuff are implementations.

bedilbek commented:
Any updates on this issue?
